remove duplicate and unecessary information printing in runtests.jl

amitmurthy opened the pull request:
I found this patch coupled with the second commit from #20276 resulting in a much saner output from runtests.jl

Largely addresses https://github.com/JuliaLang/julia/issues/18647

amitmurthy committed:
runtests.jl: remove unecessary final call stack when run in script mode

tkelman commented on test/runtests.jl:
```diff
@@ -190,7 +189,6 @@ cd(dirname(@__FILE__)) do
         println("    \033[31;1mFAILURE\033[0m")
-        Base.Test.print_test_errors(o_ts)
-        error()
+        exit(-1)
```
we probably don't want including this file from the repl to exit the calling process

amitmurthy replied:
`!isinteractive() && exit(-1)`

tkelman commented:
I've been meaning to file an issue (or check if there are any exact duplicates), but prior to the testset revamp we would actually get a line number in the test file that failed. I'm currently not seeing that, which makes debugging test failures really difficult. If we were able to get the line number back, I wouldn't want to prevent it from displaying.

amitmurthy commented:
I can see line numbers and correct call stack if a testfile fails due to an exception, but not if a `@test` fails.

With a `error("foo")` at the top of error.jl
```
Worker 2 failed running test error:
error: Error During Test
  Got an exception of type LoadError outside of a @test
  LoadError: foo
  Stacktrace:
   [1] jl_system_image_data at /Users/amitm/Julia/julia/usr/lib/julia/sys.dylib:? (repeats 17 times)
  while loading /Users/amitm/Julia/julia/test/error.jl, in expression starting on line 2
```

but with a `@test false`

```
Worker 2 failed running test error:
error: Test Failed
  Expression: false
Stacktrace:
 [1] record(::Base.Test.DefaultTestSet, ::Base.Test.Fail) at ./test.jl:564
 [2] (::##33#40)() at /Users/amitm/Julia/julia/test/runtests.jl:168
 [3] cd(::##33#40, ::String) at ./file.jl:69
 [4] include_from_node1(::String) at ./loading.jl:539
 [5] include_from_node1(::String) at /Users/amitm/Julia/julia/usr/lib/julia/sys.dylib:?
 [6] include(::String) at ./sysimg.jl:14
 [7] include(::String) at /Users/amitm/Julia/julia/usr/lib/julia/sys.dylib:?
 [8] process_options(::Base.JLOptions) at ./client.jl:304
 [9] _start() at ./client.jl:370
 [10] _start() at /Users/amitm/Julia/julia/usr/lib/julia/sys.dylib:?
```

In the second case the callstack is irrelevant.

tkelman commented:
Right, but before the testsystem change a `@test` failure would give a `while loading /.../julia/test/failurefile.jl, in expression starting on line #failurelineno#` - it was a bit needle-in-haystack, but it was there.

KristofferC commented:
As a side point I find that failed tests inside testsets have good backtraces.

tkelman commented:
but shouldn't each file have an implicit file testset now?

KristofferC commented:
Hmm, I'm haven't looked into too much how `runtests` is working so I'm just reporting how I remember things where when I last tried them.

tkelman reviewed (dismissed):
should not call exit if the file gets included

amitmurthy committed:
runtests.jl : remove duplicate printing of errors and callstacks

amitmurthy committed:
Print correct remote pid during deserialization errors

amitmurthy commented:
Fixed and split the original commit into two in case we want to revert the second one.
3rd commit fixes a `pid` value being wrongly printed (introduced in https://github.com/JuliaLang/julia/pull/20276)

Following are examples of outout in the prescence of errors with this PR.
The testsystem is hiding a lot of the original stacktraces and needs to be fixed.
In some cases the original exception is also being hidden by a TestSetException
as seen below.


Running `julia runtests.jl char env`:

With `char.jl` incorporating a dummy `@test throw(BoundsError(DontExistOn1(1), 1))`

```
Test (Worker) | Time (s) | GC (s) | GC % | Alloc (MB) | RSS (MB)
env (3)       |    0.55  |  0.00  |  0.0 | 12.12      | 211.66  
char: Error During Test
  Test threw an exception of type RemoteException
  Expression: char
  Error deserializing a remote exception from worker 2
  Remote(original) exception of type Base.Test.TestSetException
  Remote stacktrace : 
   in finish at ./test.jl:663
   in macro expansion at ./util.jl:288 [inlined]
   in macro expansion at /Users/amitm/Julia/julia/test/testdefs.jl:0 [inlined]
   in anonymous at ./<missing>:?
   in eval at ./boot.jl:236
   in runtests at /Users/amitm/Julia/julia/test/testdefs.jl:19
   in runtests at /Users/amitm/Julia/julia/test/testdefs.jl:4
   in #520 at ./multi.jl:1539
   in run_work_thunk at ./multi.jl:1085
   in macro expansion at ./multi.jl:1539 [inlined]
   in #519 at ./event.jl:73

Test Summary: | Pass  Error  Total
  Overall     |   51      1     52
    env       |   51            51
    char      |           1      1
    FAILURE
```

Same output for `throw(BoundsError(DontExistOn1(1), 1))` without a `@test`

With a `error("foo")`

```
Test (Worker) | Time (s) | GC (s) | GC % | Alloc (MB) | RSS (MB)
env (3)       |    0.57  |  0.00  |  0.0 | 12.12      | 212.23  
Worker 2 failed running test char:
char: Error During Test
  Got an exception of type LoadError outside of a @test
  LoadError: foo
  while loading /Users/amitm/Julia/julia/test/char.jl, in expression starting on line 23

Test Summary: | Pass  Error  Total
  Overall     |   64      1     65
    char      |   13      1     14
    env       |   51            51
    FAILURE
```

With a `@test 1==error("foo")`

```
Test (Worker) | Time (s) | GC (s) | GC % | Alloc (MB) | RSS (MB)
env (3)       |    0.56  |  0.00  |  0.0 | 12.12      | 213.02  
Worker 2 failed running test char:
char: Error During Test
  Test threw an exception of type ErrorException
  Expression: 1 == error("foo")
  foo

Test Summary: | Pass  Error  Total
  Overall     | 1758      1   1759
    env       |   51            51
    char      | 1707      1   1708
    FAILURE
```


With a `@test false`:
```
Test (Worker) | Time (s) | GC (s) | GC % | Alloc (MB) | RSS (MB)
env (3)       |    0.55  |  0.00  |  0.0 | 12.12      | 219.73  
Worker 2 failed running test char:
char: Test Failed
  Expression: false
Stacktrace:
 [1] record(::Base.Test.DefaultTestSet, ::Base.Test.Fail) at ./test.jl:564
 [2] (::##33#40)() at /Users/amitm/Julia/julia/test/runtests.jl:167
 [3] cd(::##33#40, ::String) at ./file.jl:69
 [4] include_from_node1(::String) at ./loading.jl:539
 [5] include(::String) at ./sysimg.jl:14
 [6] process_options(::Base.JLOptions) at ./client.jl:304
 [7] _start() at ./client.jl:370

Test Summary: | Pass  Fail  Total
  Overall     | 1758     1   1759
    env       |   51           51
    char      | 1707     1   1708
    FAILURE
```

tkelman commented on base/multi.jl:
```diff
@@ -1029,8 +1029,14 @@ function serialize(s::AbstractSerializer, ex::CapturedException)
-                                                " Original exception of type : ", ex_str))
+function original_ex(s::AbstractSerializer, ex_str, remote_stktrace)
+    local pid_str = ""
+    try pid_str = string(" from worker ", worker_id_from_socket(s.io)) end
```
make this a block instead of a one-liner

amitmurthy replied:
Why? I want to ignore any exceptions raised here and it looks pretty clear.

tkelman replied:
having `try ... end` all on the same line isn't good style for library code IMO, we often space out if, begin, for, etc for readability

amitmurthy commented:
I'll merge this tomorrow. I believe only duplicate printing has been removed. We can add back stuff if we see otherwise.

testsystem does not print backtraces consistently in all cases. But that is a an issue to be fixed in the testsystem.

tkelman commented:
The exit makes runtests uncatchable, I'm not sure that's a great idea.

Given the missing information from the testsystem, I don't think we should be actively suppressing anything more from it until we've fixed that and can be sure useful information is still shown.

amitmurthy commented:
OTOH, the `error()`, in the usual sense that `runtests.jl` is executed, is also used to exit the program but also prints an entirely irrelevant call stack. In interactive mode we are still throwing an error. We could check for an env `JULIA_TEST_EXIT_ON_FAILURE`, and exit only if not defined or set as 1 (and all common variants of "yes"), i.e., default of 1.

AFAIK, the stuff removed was printing a) the same error multiple times and b) call stacks not related the original error. IMO, once the visual noise is removed, it will actually become easier to detect when something that should be printed is not being done and correct it. The lines removed only relate to duplicate/irrelevant display of errors, not error handling. The testrun will still exit with a non-zero error code on any failure.

KristofferC commented:
The discussion about printing unnecessary stacktraces when the test runner errors was had in https://github.com/JuliaLang/julia/pull/19750. The solution there was to create a custom `Exception` type which overloads the `Base.showerror` method as to not print the backtrace.

amitmurthy commented:
I'll close this and open 2 other PRs, one to prevent irrelevant final callstack printing due to the `error()` call. And one to fix the wrong remote pid issue (3rd commit of this PR). Will drop the commenting out of reduntant printing for now.
